index.coffee

```diff
@@ -148,20 +35,171 @@ onProgressHandler = (onProgressPromise, fallbackOnProgress) ->
+
+class ProgressReporter
+	constructor: (@renderProgress) -> #
+
+	# Create a stream that transforms `docker.modem.followProgress` onProgress
+	# events to include total progress metrics.
+	pullProgress: (image, onProgress) ->
+		progressRenderer = @renderProgress
+		downloadProgressTracker = new ProgressTracker(100 * 1024) # 100 KB
+		extractionProgressTracker = new ProgressTracker(1024 * 1024) # 1 MB
+		lastPercentage = 0
+		Promise.try ->
+			return (evt) ->
+				try
+					{ id, status } = evt
+
+					if status is 'Pulling fs layer'
+						downloadProgressTracker.addLayer(id)
+						extractionProgressTracker.addLayer(id)
+					else if status is 'Downloading'
+						downloadProgressTracker.updateLayer(id, evt.progressDetail)
+					else if status is 'Extracting'
+						extractionProgressTracker.updateLayer(id, evt.progressDetail)
+					else if status is 'Download complete'
+						downloadProgressTracker.finishLayer(id)
+					else if status is 'Pull complete'
+						extractionProgressTracker.finishLayer(id)
+					else if status is 'Already exists'
+						downloadProgressTracker.finishLayer(id)
+						extractionProgressTracker.finishLayer(id)
+
+					if status.match(/^Status: Image is up to date for /) or status.match(/^Status: Downloaded newer image for /)
+						downloadedPercentage = 100
+						extractedPercentage = 100
+					else
+						downloadedPercentage = downloadProgressTracker.getProgress()
+						extractedPercentage = extractionProgressTracker.getProgress()
+
+					percentage = (downloadedPercentage + extractedPercentage) // 2
+					percentage = lastPercentage = Math.max(percentage, lastPercentage)
+
+					onProgress _.merge evt, {
+						percentage
+						downloadedPercentage
+						extractedPercentage
+						totalProgress: progressRenderer(percentage)
+					}
+				catch err
+					console.warn('Progress error:', err.message ? err)
+
+	# Create a stream that transforms `docker.modem.followProgress` onProgress
+	# events to include total progress metrics.
+	pushProgress: (image, onProgress) ->
+		progressRenderer = @renderProgress
+		progressTracker = new ProgressTracker(100 * 1024) # 100 KB
+		lastPercentage = 0
+		Promise.try ->
+			return (evt) ->
+				try
+					{ id, status } = evt
+
+					pushMatch = /Image (.*) already pushed/.exec(status)
+
+					id ?= pushMatch?[1]
+					status ?= ''
+
+					if status is 'Preparing'
+						progressTracker.addLayer(id)
+					else if status is 'Pushing' and evt.progressDetail.current?
+						progressTracker.updateLayer(id, evt.progressDetail)
+					# registry v2 statuses
+					else if _.includes(['Pushed', 'Layer already exists', 'Image already exists'], status) or /^Mounted from /.test(status)
+						progressTracker.finishLayer(id)
+					# registry v1 statuses
+					else if pushMatch? or _.includes(['Already exists', 'Image successfully pushed'], status)
+						progressTracker.finishLayer(id)
+
+					percentage = if status.match(/^latest: digest: /) or status.match(/^Pushing tag for rev /)
```

Comment:
It's probably better to use `/^.../.test(status)` or `status.startsWith('...')` in order to avoid building up the matches array we don't care about and instead just getting a boolean result